claat/fetch/fetch.go

```diff
@@ -13,8 +13,419 @@
+	body io.ReadCloser // resource body
+	mod  time.Time     // last update of content
+}
+
+// codelab wraps types.Codelab, while adding source type
+// and modified timestamp fields.
+type codelab struct {
+	*types.Codelab
+	Typ srcType   //  source type
+	Mod time.Time // last modified timestamp
+}
+
+type Fetcher struct {
+	authHelper   *auth.Helper
+	authToken    string
+	crcTable     *crc64.Table
+	passMetadata map[string]bool
+}
+
+func NewFetcher(at string, pm map[string]bool, rt http.RoundTripper) (*Fetcher, error) {
+	h, err := auth.NewHelper(at, auth.ProviderGoogle, rt)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Fetcher{
+		authHelper:   h,
+		authToken:    at,
+		crcTable:     crc64.MakeTable(crc64.ECMA),
+		passMetadata: pm,
+	}, nil
+}
+
+// SlurpCodelab retrieves and parses codelab source.
+// It takes the source, plus an auth token and a set of extra metadata to pass along.
+// It returns parsed codelab and its source type.
+//
+// The function will also fetch and parse fragments included
+// with types.ImportNode.
+func (f *Fetcher) SlurpCodelab(src string) (*codelab, error) {
+	res, err := f.fetch(src)
+	if err != nil {
+		return nil, err
+	}
+	defer res.body.Close()
+
+	opts := *parser.NewOptions()
+	opts.PassMetadata = f.passMetadata
+
+	clab, err := parser.Parse(string(res.typ), res.body, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch imports and parse them as fragments
+	var imports []*types.ImportNode
+	for _, st := range clab.Steps {
+		imports = append(imports, types.ImportNodes(st.Content.Nodes)...)
+	}
+	ch := make(chan error, len(imports))
+	defer close(ch)
+	for _, imp := range imports {
+		go func(n *types.ImportNode) {
+			frag, err := f.slurpFragment(n.URL)
+			if err != nil {
+				ch <- fmt.Errorf("%s: %v", n.URL, err)
+				return
+			}
+			n.Content.Nodes = frag
+			ch <- nil
+		}(imp)
+	}
+	for range imports {
+		if err := <-ch; err != nil {
+			return nil, err
+		}
+	}
+
+	v := &codelab{
+		Codelab: clab,
+		Typ:     res.typ,
+		Mod:     res.mod,
+	}
+	return v, nil
+}
+
+func (f *Fetcher) SlurpImages(src, dir string, steps []*types.Step) (map[string]string, error) {
+	// make sure img dir exists
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+
+	type res struct {
+		url, file string
+		err       error
+	}
+
+	ch := make(chan *res, 100)
+	defer close(ch)
+	var count int
+	for _, st := range steps {
+		nodes := types.ImageNodes(st.Content.Nodes)
+		count += len(nodes)
+		for _, n := range nodes {
+			go func(n *types.ImageNode) {
+				url := n.Src
+				file, err := f.slurpBytes(src, dir, url)
+				if err == nil {
+					n.Src = filepath.Join(util.ImgDirname, file)
+				}
+				ch <- &res{url, file, err}
+			}(n)
+		}
+	}
+
+	imap := make(map[string]string, count)
+	var err error
+	for i := 0; i < count; i++ {
+		r := <-ch
+		imap[r.file] = r.url
+		if r.err != nil && err == nil {
+			// record first error
+			err = fmt.Errorf("%s => %s: %v", r.url, r.file, r.err)
```

Comment:
Would it be helpful to aggregate all the errors?

Reply:
I think this does what we want?